core/keystore/src/lib.rs

```diff
@@ -14,185 +14,180 @@
-		let file = File::open(path)?;
+	/// Gets the key matching the public key.
+	pub fn get_key(&self, public: &TPair::Public) -> Option<TPair> {
+		let pair = {
+			let mut keys = self.keys.lock();
+			if let Some(value) = keys.get_mut(public) {
+				value.1 = Instant::now();
+				Some(value.0.clone())
+			} else {
+				None
+			}
+		};
+		self.clean();
+		pair
+	}
 
-		let phrase: String = ::serde_json::from_reader(&file)?;
-		let (pair, _) = TPair::from_phrase(&phrase, Some(password))
-			.ok().ok_or(Error::InvalidPhrase)?;
-		if &pair.public() != public {
-			return Err(Error::InvalidPassword);
-		}
-		Ok(pair)
+	/// Find keys matching a predicate.
+	pub fn get_keys(&self, f: impl Fn(&TPair::Public) -> bool) -> Vec<TPair> {
+		let filtered_keys = {
+			let mut keys = self.keys.lock();
+			let filtered_keys = keys.iter()
+				.filter(|(public, _)| f(public))
+				.map(|(_, (pair, _))| pair.clone())
+				.collect::<Vec<_>>();
+			for key in &filtered_keys {
+				keys.get_mut(&key.public())
+					.expect("We hold a lock; qed").1 = Instant::now();
+			}
+			filtered_keys
+		};
+		self.clean();
+		filtered_keys
 	}
 
-	/// Get public keys of all stored keys.
-	pub fn contents<TPublic: Public>(&self) -> Result<Vec<TPublic>> {
-		let mut public_keys: Vec<TPublic> = self.additional.keys()
-			.filter_map(|(ty, public)| {
-				if *ty != TPublic::KEY_TYPE {
-					return None
-				}
-				Some(TPublic::from_slice(public))
-			})
-			.collect();
-
-		let key_type: [u8; 4] = TPublic::KEY_TYPE.to_le_bytes();
-		for entry in fs::read_dir(&self.path)? {
-			let entry = entry?;
-			let path = entry.path();
-
-			// skip directories and non-unicode file names (hex is unicode)
-			if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
-				match hex::decode(name) {
-					Ok(ref hex) => {
-						if hex[0..4] != key_type { continue	}
-						let public = TPublic::from_slice(&hex[4..]);
-						public_keys.push(public);
-					}
-					_ => continue,
-				}
+	/// Cleanup session store.
+	fn clean(&self) {
+		let mut keys = self.keys.lock();
+		if keys.len() > 1 {
+			let duration = Duration::new(INACTIVITY_PERIOD, 0);
+			let new_keys = keys.iter().filter(|(_, (_, instant))| {
+				Instant::now().duration_since(instant.clone()) < duration
+			}).map(|(k, v)| (k.clone(), v.clone())).collect::<HashMap<_, _>>();
+			if new_keys.len() > 0 {
+				*keys = new_keys;
 			}
 		}
-
-		Ok(public_keys)
 	}
+}
 
-	fn key_file_path<TPair: Pair>(&self, public: &TPair::Public) -> PathBuf {
-		let mut buf = self.path.clone();
-		let bytes: [u8; 4] = TPair::KEY_TYPE.to_le_bytes();
-		let key_type = hex::encode(bytes);
-		let key = hex::encode(public.as_slice());
-		buf.push(key_type + key.as_str());
-		buf
+impl<TPair> StoreHandler for LocalStore<TPair>
+where
+	TPair: Pair + Send,
+	TPair::Public: Send,
+{
+	fn on_set_session_key(&self, pair: &[u8]) {
+		let pair = TPair::from_seed_slice(pair)
+			.expect("store contains valid bytes");
```

Comment:
Imho the proof is insufficient. It will only give us a correct `pair` if the handler is filtered correctly (and hence registered correctly) by it's key type. I'd rather look for a solution that passes a properly typed `Pair` here.

Reply:
Indeed. This proof is wholly insufficient. A proof should be a logical tautology and not contain assumptions about prior state.